Printer: use indentation and line feeds

agarny opened the pull request:
Addresses issue #309.

There were two ways to approach this issue:
 1) Rely on libxml2 to do the indentation and line feeding for us; and
 2) Do the indentation and line feeding manually.

I gave both approaches a try and the first approach meant creating a dummy XML document (in case we only wanted to print, say, a component since that bit of XML code might have resulted in more than one top XML element, thus making the XML code invalid for libxml2), which in turn meant getting rid of the dummy top element and unindenting its children. In other words, although it was working, it was a bit of a dirty hack. Also, because of the way we handle math, it meant that the `Parser.parseModelWithComponentsWithMultipleMathElements` test would keep failing.

So, in the end, I went with the second approach, which happens to be much cleaner. I do, however, appreciate why indentation and line feeds were not used in the first place! It's a wee bit more involved... :)

agarny committed:
Tests: some minor cleaning up.

agarny committed:
Tests: some minor cleaning up.

agarny committed:
Tests: added indentations and line feeds.

agarny committed:
Tests: added indentations and line feeds.

agarny committed:
Tests: some minor cleaning up.

Because of some previous copying/pasting.

agarny committed:
Tests: some minor cleaning up.

agarny committed:
Tests: added indentations and line feeds.

agarny committed:
Tests: some minor cleaning up.

agarny committed:
Validator: improved the integerStrings test.

agarny committed:
Printer: removed the printEncapsulation() and printWhen() methods from the private section of the Printer class.

They have nothing to do there.

agarny committed:
Parser: ignore whitespaces in a when child.

This had clearly been forgotten since this is done everywhere else.

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: improved the integerStrings test.

agarny committed:
Tests: added indentations and line feeds.

agarny committed:
Printer: added indentations and line feeds.

hsorby commented on src/xmlnode.cpp:
```diff
@@ -157,10 +157,13 @@ XmlNodePtr XmlNode::getParent()
 }
 
-std::string XmlNode::convertToString() {
+std::string XmlNode::convertToString(int format) {
```
I would have preferred an enumeration here.

agarny replied:
I thought I would mimic what `xmlNodeDump()` does, but if anything I would actually use a `bool`.

hsorby replied:
I would use `enum` if there was going to be more than two formats otherwise `bool` if there were only going to be two.  Either way not `int`.

agarny replied:
For whatever reason, libxml2 uses an `int`, but there is really only two values: `0` and `1` for no formatting and formatting, respectively.

Ok, I am going to use a `bool`.

nickerso replied:
libxml2 is a C API which has no native bool type....

hsorby commented on tests/math/math.cpp:
```diff
@@ -18,36 +18,33 @@ limitations under the License.
 
 #include <libcellml>
 
+static const std::string emptyMath = "<math xmlns=\"http://www.w3.org/1998/Math/MathML\"/>\n";
```
Shouldn't this variable be ALL CAPS?

agarny replied:
I don't mind either way. It's just that I have seen something similar in `tests/generator/generator.cpp`.

hsorby replied:
Don't look in the tests directory for code conformance!

agarny replied:
Ok, I am going to replace `emptyMath` with `EMPTY_MATH`. No problem.

hsorby reviewed (commented):
I see you have opted for 2 space indent and not 4.

agarny commented:
> I see you have opted for 2 space indent and not 4.

I initially wanted to go with a four-space indentation, but libxml2 uses two-space and I need libxml2 (through `XmlNode::convertToString()`) to format a math string, which means two-space indentation.

Hmm... I am going to check whether we cannot ask libxml2 to use four-space indentation...

hsorby commented:
I am all for going with the default that libxml2 is using that means less work adapting for what I consider zero benefit.

agarny commented:
> I am all for going with the default that libxml2 is using that means less work adapting for what I consider zero benefit.

That's why I hadn't bothered checking, but I have just seen that it's actually very easy to change the number of spaces to be used for indentation. One just needs to set `xmlTreeIndentString` to whatever (literally) we want. So, here, all we need is to have `xmlTreeIndentString = "    ";` before calling `xmlNodeDump()` in `XmlNode::convertToString()`.

I am happy to make that change, if we want...

hsorby commented:
I'm happy with two. So if there is going to be a vote I'm voting two. (I know some people are very particular about indentation :))

agarny committed:
Math test: replaced `emptyMath` with `EMPTY_MATH`.

agarny committed:
XmlNode::convertToString(): replaced the `int format` with `bool format`.

I initially wanted to mimic what libxml2 does, but in the end there are
only two possible values: 0 and 1 for no formatting and formatting,
respectively. So, we really ought to use a bool here.

agarny commented:
I prefer four myself, but I am ok with two.

nickerso commented:
This really seems to be introducing a significant maintenance burden for zero gain. If we are going to do this I would definitely prefer to simply let libXml2 to handle the "prettifying" for us.

agarny commented:
> This really seems to be introducing a significant maintenance burden for zero gain. If we are going to do this I would definitely prefer to simply let libXml2 to handle the "prettifying" for us.

I did give it a try, but this caused problems (see my original message).

hsorby commented:
> This really seems to be introducing a significant maintenance burden for zero gain. If we are going to do this I would definitely prefer to simply let libXml2 to handle the "prettifying" for us.

It's a problem either way I'm pretty sure I have already been through once (or twice) and reformatted all the expected test strings.  I'm not keen on making this prettifying change but I will be useful to be able to execute simple text diffs on the printed output from libCellML which I think will be a time saver long term.

agarny committed:
Merge branch 'develop' into issue309.

agarny committed:
Releasing Version 0.1.0.

agarny committed:
Merge remote-tracking branch 'cellml/develop' into develop.

agarny committed:
Merge branch 'develop' into issue309.

nickerso commented:
If we implement #332 then I believe the need to implement our own formatting goes away and we can simply use libXml2 to handle the code formatting, as per the original option 1.

agarny commented:
> If we implement #332 then I believe the need to implement our own formatting goes away and we can simply use libXml2 to handle the code formatting, as per the original option 1.

I agreed that option 1 should work. Still, nothing that prevents us from accepting this PR until we actually start work on issue #332.

agarny committed:
Merge remote-tracking branch 'cellml/develop' into develop.

agarny committed:
Merge branch 'develop' into issue309.
